tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -25,25 +29,189 @@ pub struct CudaVec<T: Numeric> {
 }
 
 impl<T: Numeric> CudaVec<T> {
+    pub fn new(len: usize, stream: &CudaStream) -> Self {
+        let vec = unsafe { Self::new_async(len, stream) };
+        stream.synchronize();
+        vec
+    }
     /// # Safety
     ///
-    /// - `ptr` __must__ be a valid device pointer to an array of `len` elements of type `T`
-    pub unsafe fn new(ptr: CudaPtr, len: usize, device: CudaDevice) -> Self {
+    /// - `stream` __must__ be synchronized to guarantee computation has finished
+    pub unsafe fn new_async(len: usize, stream: &CudaStream) -> Self {
+        let size = len as u64 * std::mem::size_of::<T>() as u64;
+        let ptr = CudaPtr {
+            ptr: cuda_malloc_async(size, stream.as_c_ptr()),
+            device: stream.device,
+        };
+
         Self {
             ptr,
             len,
-            device,
+            device: stream.device,
             _phantom: PhantomData,
         }
     }
+
```

Comment:
Here, does cuda_malloc initialize memory ?

If it does not, I think we should memset it as, uninitialized memory is unsafe

Reply:
Makes sense, will do